### PR TITLE
🌱 Bump golang to v1.21.6

### PR DESCRIPTION
**Pull request by kashifest**
as the title says

### CODE REVIEW

**Comment by kashifest** [comment]
/hold
this requires project-infra patch as well

**Comment by smoshiur1237** [comment]
Is this PR needs the project-infra PR to merge first. The gomod test is failing because of new go version.

**Review by tuminoid — COMMENTED** [review]
```
hack/ensure-go.sh
35:    minimum_go_version=go1.20
```
Missed this, otherwise LGTM.

**Comment by tuminoid** [comment]
> Is this PR needs the project-infra PR to merge first? The gomod test is failing because of new go version.

Project infra needs to go first whenever Prow test is broken/changed.

**Comment by kashifest** [comment]
> ```
> hack/ensure-go.sh
> 35:    minimum_go_version=go1.20
> ```
> 
> Missed this, otherwise LGTM.

Intentionally left out since the PR job was failing, will bump it in a followup

**Review by tuminoid — COMMENTED** [review]
/lgtm 

Tests will start passing when project-infra one is merged.

File: controllers/metal3data_controller_test.go [hunk 1]
```diff
@@ -80,11 +80,11 @@ var _ = Describe("Metal3Data manager", func() {
 				}
 				if tc.m3d != nil && !tc.m3d.DeletionTimestamp.IsZero() {
 					if tc.releaseLeasesRequeue {
-						m.EXPECT().ReleaseLeases(context.TODO()).Return(baremetal.WithTransientError(errors.New(""), requeueAfter))
+						m.EXPECT().ReleaseLeases(context.Background()).Return(baremetal.WithTransientError(errors.New(""), requeueAfter))
```
Comment (mboukhalfa) [inline controllers/metal3data_controller_test.go:83]: was changing the context from todo to background required by the new go version ?

Comment (kashifest) [reply controllers/metal3data_controller_test.go:83]: Yes, the function uses context background but the test was expecting context todo

**Pull request merged**